### ISSUE TITLE
Invoking a callback protocol with variadic arguments fails to type check

### ISSUE THREAD

**Issue by chris-chambers** [microsoft/pyright#2950]
**Describe the bug**
Related to #2948, invoking a callback protocol containing variadic arguments with a matching unpacked TypeVarTuple fails to typecheck.

**To Reproduce**
Define such a callback protocol (see attached code) and try to invoke it.

**Expected behavior**
The call site should typecheck without error.

**Screenshots or Code**
```python
# pyright: strict

from typing import Protocol, TypeVar
from typing_extensions import TypeVarTuple, Unpack


T = TypeVar("T")
Ts = TypeVarTuple("Ts")


class CallbackPosOnly(Protocol[Unpack[Ts]]):
    def __call__(self, *args: *Ts) -> tuple[Unpack[Ts]]:
        ...


def invoke_posonly(fn: CallbackPosOnly[Unpack[Ts]], *args: *Ts) -> tuple[Unpack[Ts]]:
    return fn(*args)


class CallbackKeyed(Protocol[Unpack[Ts], T]):
    def __call__(self, *args: *Ts, keyed: T) -> tuple[Unpack[Ts], T]:
        ...


def invoke_keyed(fn: CallbackKeyed[Unpack[Ts], T], *args: *Ts) -> tuple[Unpack[Ts], T]:
    return fn(*args, keyed=True)
```

Pyright output:

```
No configuration file found.
No pyproject.toml file found.
stubPath /Users/chris/projects/pyright-bugs/2/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Auto-excluding /Users/chris/projects/pyright-bugs/2/venv
Found 1 source file
/Users/chris/projects/pyright-bugs/2/example.py
  /Users/chris/projects/pyright-bugs/2/example.py:17:16 - error: Unpacked argument cannot be used for TupleTypeVar parameter (reportGeneralTypeIssues)
  /Users/chris/projects/pyright-bugs/2/example.py:17:12 - error: Return type is unknown (reportUnknownVariableType)
  /Users/chris/projects/pyright-bugs/2/example.py:26:16 - error: Unpacked argument cannot be used for TupleTypeVar parameter (reportGeneralTypeIssues)
  /Users/chris/projects/pyright-bugs/2/example.py:26:12 - error: Return type is unknown (reportUnknownVariableType)
4 errors, 0 warnings, 0 informations 
Completed in 0.536sec
```

**VS Code extension or command-line**
Present in both:

- Pyright 1.1.216
- Pylance 2022.1.5

**Comment by erictraut** [comment]
Thanks for the bug report.

The error with the `invoke_posonly` is a valid bug, and it will be fixed in the next release.

Even with the bug fix in place, the `invoke_keyed` sample will not type check, however. This has nothing to do with variadic type variables. The following example also does not type check. Both pyright and mypy generate errors in this case.

```python
class CallbackKeyed(Protocol[T]):
    def __call__(self, keyed: T) -> T: ...

def invoke_keyed_protocol(fn: CallbackKeyed[T]) -> T:
    return fn(True) # Error

def invoke_keyed_callable(fn: Callable[[T], T]) -> T:
    return fn(True) # Error
```

**Comment by chris-chambers** [comment]
Sorry about the noise on the keyed protocol.  I was reducing the sample from a larger example, and removed too much for it to be valid.  I don't think the `keyed` parameter will have an effect on the underlying problem, but I wanted to include a keyword-only param for completeness.

The code should have been more like:

```python
class CallbackKeyed(Protocol[Unpack[Ts]]):
    def __call__(self, *args: *Ts, keyed: bool) -> tuple[Unpack[Ts]]:
        ...


def invoke_keyed(fn: CallbackKeyed[Unpack[Ts]], *args: *Ts) -> tuple[Unpack[Ts]]:
    return fn(*args, keyed=True)


def invoke_keyed_should_fail(fn: CallbackKeyed[Unpack[Ts]], *args: *Ts) -> tuple[Unpack[Ts]]:
    # keyed should understood as keyword-only, and the following line should not typecheck
    return fn(*args, True)
```

**Comment by erictraut** [comment]
Thanks, I've included your sample as part of the unit test. This will be fixed in the next release.

**Comment by chris-chambers** [comment]
Thanks again for the quick resolution, and thanks for your work on Pyright.  It's an excellent tool.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.217, which I just published. It will also be included in the next release of pylance.
